pkg/kv/kvserver/kvflowcontrol/replica_rac2/admission.go

```diff
@@ -0,0 +1,277 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package replica_rac2
+
+import (
+	"cmp"
+	"slices"
+
+	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
+	"github.com/cockroachdb/errors"
+)
+
+// lowPriOverrideState records which raft log entries have their priority
+// overridden to be raftpb.LowPri. Used at follower replicas.
+//
+// The lifetime of a particular log index begins when the entry is appended
+// to the in-memory log, and ends when the entry is sent to storage with the
+// priority that is extracted from this struct. The same log index can
+// reappear in this struct multiple times, when a new raft leader overwrites a
+// suffix of the log proposed by the previous leader, or if the existing
+// leader retransmits an entry.
+//
+// sideChannelForLowPriOverride is called when stepping the MsgApps in a
+// RaftMessageRequest, and getEffectivePriority is called from within
+// handleRaftReady when processing the corresponding MsgStorageAppend.
+//
+// Maintaining accuracy in the presence of an existing leader doing retransmits
+// is tricky, since the caller does not know which MsgApps were accepted by
+// Raft during a step. We can have two kinds of situations:
+//
+//   - [C1] e10 was stepped, but not accepted, since e9 was missing. e10 is
+//     stepped successfully later.
+//
+//   - [C2] e10 was successfully stepped, but the leader did not realize it, and
+//     later resends e10 that is ignored.
+//
+// We specify a weak correctness requirement: if a leader has not overridden
+// to LowPri, the lowPriOverrideState should not think it has. This prevents a
+// situation where an entry is consuming regular tokens on the leader, but
+// will be admitted as elastic work on the follower.
+//
+// That is, we allow for the following error: the leader has overridden to
+// LowPri, but lowPriOverrideState thinks it has not.
+//
+// RACv2 has an invariant that if a leader uses LowPri override once for an
+// entry, it must be in the send-queue, and will never again be used without
+// an override. This gives us an easy way to achieve the correctness
+// requirement -- use the specification of override (or lack thereof) from the
+// first time that a leader provides us information for an index, and don't
+// change it.
+//
+// For C2, consider the following sequence:
+// - e10 NormalPri stepped successfully
+// - e10 LowPri override not stepped successfully
+// - e10 emerges in handleRaftReady: we will use NormalPri
+//
+// For C1, consider the following sequence:
+//   - e10 NormalPri not stepped successfully
+//   - e10 LowPri override stepped successfully
+//   - e10 emerges in handleRaftReady: we will use NormalPri, though should have
+//     used LowPri.
+//
+// TODO(sumeer): consider increasing the accuracy by returning information
+// from step.
+type lowPriOverrideState struct {
```

Comment:
Do we need this for the no-send-queue phase? Suggest factoring it out from this PR.